parquet-variant-compute/src/shred_variant.rs

```diff
@@ -0,0 +1,717 @@
+use arrow::compute::CastOptions;
+use arrow::datatypes::{DataType, Fields};
+use arrow::error::{ArrowError, Result};
+use parquet_variant::{ObjectBuilder, ReadOnlyMetadataBuilder, Variant, EMPTY_VARIANT_METADATA};
+
+use indexmap::IndexMap;
+use std::sync::Arc;
+
+/// Shreds the input binary variant using a target shredding schema derived from the requested data type.
+///
+/// For example, requesting `DataType::Int64` would produce an output variant array with the schema:
+///
+/// ```
+/// {
+///    metadata: BINARY,
+///    value: BINARY,
+///    typed_value: LONG,
+/// }
+/// ```
+///
+/// Similarly, requesting `DataType::Struct` with two integer fields `a` and `b` would produce an
+/// output variant array with the schema:
+///
+/// ```
+/// {
+///   metadata: BINARY,
+///   value: BINARY,
+///   typed_value: {
+///     a: {
+///       value: BINARY,
+///       typed_value: INT,
+///     },
+///     b: {
+///       value: BINARY,
+///       typed_value: INT,
+///     },
+///   }
+/// }
+/// ```
+pub fn shred_variant(array: &VariantArray, as_type: &DataType) -> Result<VariantArray> {
+    if array.typed_value_field().is_some() {
+        return Err(ArrowError::InvalidArgumentError(
+            "Input is already shredded".to_string(),
+        ));
+    }
+
+    if array.value_field().is_none() {
+        // all-null case
+        return Ok(VariantArray::from_parts(
+            array.metadata_field().clone(),
+            None,
+            None,
+            None,
+        ));
+    };
+
+    let cast_options = CastOptions::default();
+    let mut builder = make_variant_to_shredded_variant_arrow_row_builder(
+        as_type,
+        &cast_options,
+        array.len(),
+        true,
+    )?;
+    for i in 0..array.len() {
+        if array.is_null(i) {
+            builder.append_null()?;
+        } else {
+            builder.append_value(array.value(i))?;
+        }
+    }
+    let (value, typed_value, nulls) = builder.finish()?;
+    Ok(VariantArray::from_parts(
+        array.metadata_field().clone(),
+        Some(value),
+        Some(typed_value),
+        nulls,
+    ))
+}
+
+pub(crate) fn make_variant_to_shredded_variant_arrow_row_builder<'a>(
+    data_type: &'a DataType,
+    cast_options: &'a CastOptions,
+    capacity: usize,
+    top_level: bool,
+) -> Result<VariantToShreddedVariantRowBuilder<'a>> {
+    let builder = match data_type {
+        DataType::Struct(fields) => {
+            let typed_value_builder = VariantToShreddedObjectVariantRowBuilder::try_new(
+                fields,
+                cast_options,
+                capacity,
+                top_level,
+            )?;
+            VariantToShreddedVariantRowBuilder::Object(typed_value_builder)
+        }
+        DataType::List(_)
+        | DataType::LargeList(_)
+        | DataType::ListView(_)
+        | DataType::LargeListView(_)
+        | DataType::FixedSizeList(..) => {
+            // TODO: Special handling for shredded variant arrays
```

Comment:
this isn't really about shredded variant arrays, right? It is more like

```suggestion
            // TODO: handling for structured arrays
```

Reply:
Not sure I follow? Or maybe we're just respectively looking at it from input vs. output perspective?

After this PR, we will be able to shred primitive variant values (into primitive arrays) and object variant values (into struct arrays), but not (yet) shred array variant values (into one of the various list array types).

Reply:
Oh, I think the wording might be confusing, I'll adjust